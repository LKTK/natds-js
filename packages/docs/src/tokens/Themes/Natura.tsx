import * as React from 'react';
import { themes } from '@naturacosmeticos/natds-styles';
import Table from '@addons/Table/Table';

const { natura } = themes;

function FlattenTheme(accum: any, [key, value]: any): any {
  const type = typeof value;

  if (type === 'function' || type === 'object' && !!value) {
    return Object.entries(value).reduce(FlattenTheme, [key, ...accum]);
  }

  const [parent] = accum;

  if (!parent) return accum;

  const name = `${parent}-${key}`;

  return [...accum, [{ name }, { name: value, type: 'color' }]];
}

function removeStrings(entry: any) {
  return typeof entry !== 'string';
}

const { light, dark } = natura;

const darkSchema = {
  head: ['name', 'color'],
  body: (dark && dark.palette) && Object.entries(dark.palette).reduce(FlattenTheme, []).filter(removeStrings)
};

darkSchema.body.push(
  [
    {
      name: 'webFontFamilyBranding',
      type: 'text'
    },
    {
      name: dark.typography.web.fontFamilyBrand,
      type: 'text'
    }
  ],
  [
    {
      name: 'androidFontFamilyBranding',
      type: 'text'
    },
    {
      name: dark.typography.android.fontFamilyBrand,
      type: 'text'
    }
  ],
  [
    {
      name: 'iosFontFamilyBranding',
      type: 'text'
    },
    {
      name: dark.typography.iOS.fontFamilyBrand,
      type: 'text'
    }
  ]
);

const lightSchema = {
  head: ['name', 'color'],
  body: (light && light.palette) && Object.entries(light.palette).reduce(FlattenTheme, []).filter(removeStrings)
};

lightSchema.body.push(
  [
    {
      name: 'webFontFamilyBranding',
      type: 'text'
    },
    {
      name: light.typography.web.fontFamilyBrand,
      type: 'text'
    }
  ],
  [
    {
      name: 'androidFontFamilyBranding',
      type: 'text'
    },
    {
      name: light.typography.android.fontFamilyBrand,
      type: 'text'
    }
  ],
  [
    {
      name: 'iosFontFamilyBranding',
      type: 'text'
    },
    {
      name: light.typography.iOS.fontFamilyBrand,
      type: 'text'
    }
  ]
);

export const NaturaLightComponent = () => <Table schema={lightSchema} title="Light" />;
export const NaturaDarkComponent = () => <Table schema={darkSchema} title="Dark" />;
