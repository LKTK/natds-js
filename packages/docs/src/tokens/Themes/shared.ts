function flattenTheme(accum: any, [key, value]: any): any {
  const type = typeof value;

  if (type === 'function' || (type === 'object' && !!value)) {
    return Object.entries(value).reduce(flattenTheme, [key, ...accum]);
  }

  const [parent] = accum;

  if (!parent) return accum;

  const name = `${parent}-${key}`;

  return [...accum, [{ name }, { name: value, type: 'color' }]];
}

function removeStrings(entry: any) {
  return typeof entry !== 'string';
}

export function parseTheme(theme: any) {
  const schema = {
    head: ['name', 'color'],
    body:
      theme &&
      theme.palette &&
      Object.entries(theme.palette)
        .reduce(flattenTheme, [])
        .filter(removeStrings)
  };

  schema.body.push(
    [
      {
        name: 'fontFamilyBranding1',
        type: 'text',
      },
      {
        name: theme.typography.fontFamilyBrand1
          ? theme.typography.fontFamilyBrand1
          : 'Undefined',
        type: 'text',
      },
    ],
    [
      {
        name: 'fontFamilyBranding2',
        type: 'text',
      },
      {
        name: theme.typography.fontFamilyBrand2
          ? theme.typography.fontFamilyBrand2
          : 'Undefined',
        type: 'text',
      },
    ],
    [
      {
        name: 'fontFamilyBranding3',
        type: 'text',
      },
      {
        name: theme.typography.fontFamilyBrand3
          ? theme.typography.fontFamilyBrand3
          : 'Undefined',
        type: 'text',
      },
    ],
  );

  return schema;
}
