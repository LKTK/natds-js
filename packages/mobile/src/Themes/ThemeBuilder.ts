import { IThemeShape } from '../Provider/IThemeShape';
import { DefaultTheme } from 'react-native-paper';
import { ITheme } from '@naturacosmeticos/natds-styles';

export function ThemeBuilder(theme: ITheme | undefined, defaultTheme?: ITheme) {
  if (!theme && !defaultTheme) {
    return {};
  }

  const provider = theme ? theme : (defaultTheme || {});
  const { palette = {}, shape } = provider;

  const {
    primary: { main: primary } = { main: undefined },
    secondary: { main: secondary } = { main: undefined },
    error: { main: error } = { main: undefined },
    background = {},
    text = {}
  } = palette;

  const colors = {
    primary,
    accent: secondary,
    secondary,
    error,
    background: background.default,
    surface: background.paper,
    text: text.primary,
    onBackground: text.primary,
    onSurface: text.secondary,
    disabled: text.disabled,
    placeholder: text.hint
  };

  const roundness = shape && shape.borderRadius
    ? shape.borderRadius
    : DefaultTheme.roundness;

  const isDarkMode = palette.type === 'dark';

  const newTheme: IThemeShape = {
    ...DefaultTheme,
    dark: isDarkMode,
    roundness,
    colors: {
      ...DefaultTheme.colors,
      ...colors
    }
  };

  return newTheme;
}
