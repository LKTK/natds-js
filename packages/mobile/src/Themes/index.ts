export { themes } from '@naturacosmeticos/natds-styles';
