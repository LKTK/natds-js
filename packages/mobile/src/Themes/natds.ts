import { themes } from '@naturacosmeticos/natds-styles';
import { DefaultTheme, ThemeShape } from 'react-native-paper';

const { natds } = themes;

const natdsTheme: ThemeShape =  {
  ...DefaultTheme,
  ...natds
};

export default natdsTheme;
