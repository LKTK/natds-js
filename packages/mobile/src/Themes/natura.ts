import { themes, tokens } from '@naturacosmeticos/natds-styles';
import { DefaultTheme } from 'react-native-paper';
import { IThemeShape } from 'Provider/IThemeShape';

const { natura: { light } } = themes;
const { spacing } = tokens.spacing;

const natdsTheme: IThemeShape = {
  ...DefaultTheme,
  ...light,
  spacing
};

export default natdsTheme;
