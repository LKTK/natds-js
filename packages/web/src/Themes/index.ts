import {
  themes as styleThemes,
} from "@naturacosmeticos/natds-styles";
import { parseTheme } from "./parseTheme";
import { ThemeBrandName } from "./ThemeBrandName";
import createThemesObject from "./createThemesObject";
import { IThemesWeb } from "./IThemesWeb";

export { IThemeWeb } from "./IThemeWeb";

export const themes = Object.keys(styleThemes).reduce((value: IThemesWeb<ThemeBrandName>, key: string) => {

  const result = value;

  result[key as ThemeBrandName].dark = parseTheme(styleThemes[key as ThemeBrandName].dark);
  result[key as ThemeBrandName].light = parseTheme(styleThemes[key as ThemeBrandName].light);

  return result;
}, createThemesObject());
