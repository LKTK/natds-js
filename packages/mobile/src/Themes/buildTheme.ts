import { IThemeShape } from '../Provider/IThemeShape';
import { DefaultTheme } from 'react-native-paper';
import { IThemeMobile } from 'Themes';

export function buildTheme(theme: IThemeMobile | undefined, defaultTheme: IThemeMobile): IThemeShape {
  const provider = theme ? theme : defaultTheme;
  const { palette = {}, shape } = provider;

  const {
    primary: { main: primary } = { main: undefined },
    secondary: { main: secondary } = { main: undefined },
    error: { main: error } = { main: undefined },
    background = {},
    text = {}
  } = palette;

  const defaultColors = DefaultTheme.colors;

  const colors = {
    primary: primary ? primary : defaultColors.primary,
    accent: secondary ? secondary : defaultColors.accent,
    secondary,
    error: error ? error : defaultColors.error,
    background: background.default ? background.default : defaultColors.backdrop,
    surface: background.paper ? background.paper : defaultColors.surface,
    text: text.primary ? text.primary : defaultColors.text,
    onBackground: text.primary ? text.primary : defaultColors.onBackground,
    onSurface: text.secondary ? text.secondary : defaultColors.onSurface,
    disabled: text.disabled ? text.disabled : defaultColors.disabled,
    placeholder: text.hint ? text.hint : defaultColors.placeholder
  };

  const roundness = shape && shape.borderRadius
    ? shape.borderRadius
    : DefaultTheme.roundness;

  const isDarkMode = palette.type === 'dark';

  const newTheme: IThemeShape = {
    ...DefaultTheme,
    dark: isDarkMode,
    roundness,
    colors: {
      ...DefaultTheme.colors,
      ...colors
    }
  };

  return newTheme;
}
