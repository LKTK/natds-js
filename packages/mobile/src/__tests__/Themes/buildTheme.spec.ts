import { buildTheme } from '../../Themes/buildTheme';
import { themes, IThemeMobile } from '../../Themes';
// tslint:disable-next-line:no-var-requires
const parsedNaturaLightTheme = require('./parsedNaturaLightTheme.json');
// tslint:disable-next-line:no-var-requires
const parsedNaturaLightThemeWithoutValues = require('./parsedNaturaLightThemeWithoutValues.json');

describe('Function buildTheme', () => {
  test('it should return the parsed natura light theme', () => {
    const { natura: { light } } = themes;

    const theme = buildTheme(light, light);
    expect(theme).toStrictEqual(parsedNaturaLightTheme);
  });

  test('it should return the parsed natura light theme with missing values', () => {
    const { natura: { light } } = themes;

    const themeWithMissingValues:IThemeMobile = {
      ...light,
      shape: undefined,
      palette: undefined,
      typography: {
        fontFamily: undefined,
        fontFamilyBrand: undefined,
        fontWeightLight: undefined,
        fontWeightRegular: undefined,
        fontWeightMedium: undefined,
        fontWeightBold: undefined
      }
    };

    expect(buildTheme(themeWithMissingValues, light)).toStrictEqual(parsedNaturaLightThemeWithoutValues);
  });
});
