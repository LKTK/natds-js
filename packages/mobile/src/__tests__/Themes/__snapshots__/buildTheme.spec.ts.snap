// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function buildTheme it should return the parsed natura light theme with missing values: Natura Light Theme With Missing Values 1`] = `
Object {
  "animation": Object {
    "scale": 1,
  },
  "avatarSizes": Object {
    "huge": Object {
      "fontSize": "1.5rem",
      "size": 80,
    },
    "large": Object {
      "fontSize": "1rem",
      "size": 48,
    },
    "small": Object {
      "fontSize": ".875rem",
      "size": 32,
    },
    "standard": Object {
      "fontSize": "1rem",
      "size": 40,
    },
    "tiny": Object {
      "fontSize": ".625rem",
      "size": 24,
    },
  },
  "badgeBorderRadius": 100,
  "buttonSizes": Object {
    "large": Object {
      "minHeight": 48,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
    },
    "medium": Object {
      "minHeight": 40,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 12,
    },
    "small": Object {
      "minHeight": 32,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 8,
    },
  },
  "colors": Object {
    "accent": "#03dac4",
    "backdrop": "rgba(0, 0, 0, 0.5)",
    "background": "rgba(0, 0, 0, 0.5)",
    "disabled": "rgba(0, 0, 0, 0.26)",
    "error": "#B00020",
    "link": "#227BBD",
    "notification": "#f50057",
    "onBackground": "#000000",
    "onSurface": "#000000",
    "placeholder": "rgba(0, 0, 0, 0.54)",
    "primary": "#6200ee",
    "secondary": "#03dac4",
    "success": "#569A32",
    "surface": "#ffffff",
    "text": "#000000",
    "textHint": "#BBBBBB",
    "textSecondary": "#777777",
    "warning": "#FCC433",
  },
  "dark": false,
  "fonts": Object {
    "light": Object {
      "fontFamily": "System",
      "fontWeight": "300",
    },
    "medium": Object {
      "fontFamily": "System",
      "fontWeight": "500",
    },
    "regular": Object {
      "fontFamily": "System",
      "fontWeight": "400",
    },
    "thin": Object {
      "fontFamily": "System",
      "fontWeight": "100",
    },
  },
  "roundness": 4,
  "spacing": 8,
  "typography": Object {
    "body1": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 16,
      "fontWeight": 400,
    },
    "body2": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 14,
      "fontWeight": 400,
    },
    "button": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 14,
      "fontWeight": 600,
      "textTransform": "uppercase",
    },
    "caption": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 12,
      "fontWeight": 400,
    },
    "h1": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 96,
      "fontWeight": 300,
    },
    "h2": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 60,
      "fontWeight": 300,
    },
    "h3": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 48,
      "fontWeight": 400,
    },
    "h4": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 34,
      "fontWeight": 400,
    },
    "h5": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 24,
      "fontWeight": 400,
    },
    "h6": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 20,
      "fontWeight": 600,
    },
    "overline": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 12,
      "fontWeight": 400,
      "textTransform": "uppercase",
    },
    "subtitle1": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 16,
      "fontWeight": 400,
    },
    "subtitle2": Object {
      "fontFamily": "\\"System\\", sans-serif",
      "fontSize": 14,
      "fontWeight": 500,
    },
  },
}
`;

exports[`Function buildTheme it should return the parsed natura light theme: Natura Light Theme 1`] = `
Object {
  "animation": Object {
    "scale": 1,
  },
  "avatarSizes": Object {
    "huge": Object {
      "fontSize": "1.5rem",
      "size": 80,
    },
    "large": Object {
      "fontSize": "1rem",
      "size": 48,
    },
    "small": Object {
      "fontSize": ".875rem",
      "size": 32,
    },
    "standard": Object {
      "fontSize": "1rem",
      "size": 40,
    },
    "tiny": Object {
      "fontSize": ".625rem",
      "size": 24,
    },
  },
  "badgeBorderRadius": 100,
  "buttonSizes": Object {
    "large": Object {
      "minHeight": 48,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
    },
    "medium": Object {
      "minHeight": 40,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 12,
    },
    "small": Object {
      "minHeight": 32,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 8,
    },
  },
  "colors": Object {
    "accent": "#FF6B0B",
    "backdrop": "rgba(0, 0, 0, 0.5)",
    "background": "#FFFFFF",
    "disabled": "#EEEEEE",
    "error": "#E74627",
    "link": "#227BBD",
    "notification": "#f50057",
    "onBackground": "#333333",
    "onSurface": "#777777",
    "placeholder": "#BBBBBB",
    "primary": "#F4AB34",
    "secondary": "#FF6B0B",
    "success": "#569A32",
    "surface": "#FAFAFA",
    "text": "#333333",
    "textHint": "#BBBBBB",
    "textSecondary": "#777777",
    "warning": "#FCC433",
  },
  "dark": false,
  "fonts": Object {
    "light": Object {
      "fontFamily": "System",
      "fontWeight": "300",
    },
    "medium": Object {
      "fontFamily": "System",
      "fontWeight": "500",
    },
    "regular": Object {
      "fontFamily": "System",
      "fontWeight": "400",
    },
    "thin": Object {
      "fontFamily": "System",
      "fontWeight": "300",
    },
  },
  "roundness": 4,
  "spacing": 8,
  "typography": Object {
    "body1": Object {
      "fontFamily": "System",
      "fontSize": 16,
      "fontWeight": 400,
    },
    "body2": Object {
      "fontFamily": "System",
      "fontSize": 14,
      "fontWeight": 400,
    },
    "button": Object {
      "fontFamily": "System",
      "fontSize": 14,
      "fontWeight": 600,
      "textTransform": "uppercase",
    },
    "caption": Object {
      "fontFamily": "System",
      "fontSize": 12,
      "fontWeight": 400,
    },
    "h1": Object {
      "fontFamily": "System",
      "fontSize": 96,
      "fontWeight": 300,
    },
    "h2": Object {
      "fontFamily": "System",
      "fontSize": 60,
      "fontWeight": 300,
    },
    "h3": Object {
      "fontFamily": "System",
      "fontSize": 48,
      "fontWeight": 400,
    },
    "h4": Object {
      "fontFamily": "System",
      "fontSize": 34,
      "fontWeight": 400,
    },
    "h5": Object {
      "fontFamily": "System",
      "fontSize": 24,
      "fontWeight": 400,
    },
    "h6": Object {
      "fontFamily": "System",
      "fontSize": 20,
      "fontWeight": 600,
    },
    "overline": Object {
      "fontFamily": "System",
      "fontSize": 12,
      "fontWeight": 400,
      "textTransform": "uppercase",
    },
    "subtitle1": Object {
      "fontFamily": "System",
      "fontSize": 16,
      "fontWeight": 400,
    },
    "subtitle2": Object {
      "fontFamily": "System",
      "fontSize": 14,
      "fontWeight": 500,
    },
  },
}
`;
